File: rubyFiles/cell.rb
class Cell

    attr_reader :row, :column
    attr_accessor :north, :south, :east, :west

    #constructor
    def initialize(row, column)
        @row, @column = row, column
        @links = {}
    end

    #links two cells
    def link(cell, bidi=true)
        @links[cell] = true
        cell.link(self, false) if bidi
        self
    end

    #unlinks two cells
    def unlink(cell, bidi=true)
        @links.delete(cell)
        cell.unlink(self, false) if bidi
        self
    end

    #gets all cells linked to this one
    def links
        @links.keys
    end

    #sees if cell is linked to another cell
    def linked?(cell)
        @linke.key?(cell)
    end

    def neighbrs
        list = []
        list << north if north
        list << south if south
        list << east if east
        list << west if west
    end

end

File: pythFiles/cell.py
from distances import Distances

class Cell:

    def __init__(self, row, column):
        self.row = row
        self.column = column
        self.links = {} 
        self.north = None
        self.south = None
        self.east = None
        self.west = None

    def link(self, cell, bidi=True):
        self.links[cell] = True
        if bidi:
            cell.link(self, False)
        return self
    
    def unlink(self, cell, bidi=True):
        self.links.remove(cell)
        if bidi:
            cell.unlink(self, False)
        return self

    def links(self):
        return self.link.keys()

    def is_linked(self, cell):
        return cell in self.links

    def neighbors(self):
        neighbor = []
        if self.north is not None:
            neighbor.append(self.north)
        if self.south is not None:
            neighbor.append(self.south)
        if self.east is not None:
            neighbor.append(self.east)
        if self.west is not None:
            neighbor.append(self.west)

        return neighbor
    
    def distances(self):
        distances = Distances(self)
        frontier = []
        frontier.append(self)

        while frontier:
            new_frontier = []

            for cell in frontier:
                for link in cell.links:
                    if distances.keyExists(link) == False:
                        distances[link] = distances[cell]+1
                        new_frontier.append(link)
            frontier = new_frontier
        return distances


